certloader: Fix reload on K8s Secret/ConfigMap update

gandro opened the pull request:
This change fixes the certloader Watch function to use the new fswatcher
package which supports the event sequence emitted when Kubernetes 
updates a volume mount for a ConfigMap or Secret.

The previous code did not work as expected on Kubernetes. Kubernetes
sets up a volume mount with symlinks containing a double indirection,
see e.g. spf13/viper#415 (comment)

Simply adding the certificate paths to fsnotify therefore is not
sufficient, as fsnotify watches the symlink target, which gets removed
once Kubernetes updates the volume. Therefore the certloader code would
handle a single update before its fsnotify watches became stale and it
would stop picking up new changes.

This change uses the newly introduced fswatcher package instead. It
wraps fsnotify, but is designed to work with non-existing paths and
symlink structures which contain a double indirection.

See the code comments for details:
https://github.com/cilium/cilium/blob/cda1f045b7489b058cb1fa00e741f8a70d060105/pkg/crypto/certloader/fswatcher/fswatcher.go#L36-L61


Fixes: #13621

aanm commented on pkg/crypto/certloader/watcher.go:
```diff
@@ -118,61 +125,59 @@ func (w *Watcher) Watch() <-chan struct{} {
+				// also coalesces).
+				if reload == nil {
+					reload = time.After(watcherCoalesceWindow)
 				}
```
am I reading this correctly? doesn't this mean that this is essentially polling?

gandro replied:
Not really. The current approach suffers from spurious unrelated events causing unnecessary reloads, but it's still very different from polling. Note that the reload is only triggered when there is an event, not on a timer (which is what polling would do).

But I put the PR back into draft mode, there might be a more precise way to deal with the K8s sequence.

gandro replied:
Resolved as outdated. The new approach does not suffer from spurious events anymore.

gandro committed:
certloader: Add fswatcher package

This package adds a wrapper around fsnotify to support watching for
files which do not exist yet. See the inline comments for more details
for how it is implemented and its limitations.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
certloader: Fix reload on K8s Secret/ConfigMap update

This change fixes the certloader Watch function to use the new fswatcher
package which supports the event sequence emitted when Kubernetes
updates a volume mount for a ConfigMap or Secret.

The previous code did not work as expected on Kubernetes. Kubernetes
sets up a volume mount with symlinks containing a double indirection,
see e.g. https://github.com/spf13/viper/pull/415#issuecomment-344050279

Simply adding the certificate paths to fsnotify therefore is not
sufficient, as fsnotify watches the symlink target, which gets removed
once Kubernetes updates the volume. Therefore the certloader code would
handle a single update before its fsnotify watches became stale and it
would stop picking up new changes.

This change uses the newly introduced fswatcher package instead. It
wraps fsnotify, but is designed to work with non-existing paths and
symlink structures which contain a double indirection.

Fixes: #13621

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
certloader: Implement event coalescing

When the certificate files are mounted in a K8s volume, it is common to
observe a sequence of CHMOD, REMOVE, CREATE for each file. This causes
the files being reloaded three times (once for each event).

This commit introduces event coalescing, meaning that we delay the
reload for a short amount of time. This means if there are additional
events which would also cause a reload within the time frame, these
additional events are coalesced with the inital event.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

christarazi commented on pkg/crypto/certloader/fswatcher/fswatcher.go:
```diff
@@ -0,0 +1,356 @@
+import (
+	"errors"
+	"fmt"
+	"github.com/sirupsen/logrus"
```
Nit: import ordering

christarazi commented on pkg/crypto/certloader/fswatcher/fswatcher_test.go:
```diff
@@ -0,0 +1,160 @@
+package fswatcher
+
+import (
+	. "gopkg.in/check.v1"
```
Nit: import ordering

christarazi replied:
You might have missed this one?

gandro replied:
Argh, yes! I must have messed up my rebase somehow (I did fix it, but then apparently commited the wrong file). Should be fixed now.

christarazi commented on pkg/crypto/certloader/fswatcher/fswatcher.go:
```diff
@@ -26,6 +25,7 @@ import (
 	"github.com/cilium/cilium/pkg/logging/logfields"
 
 	"github.com/fsnotify/fsnotify"
+	"github.com/sirupsen/logrus"
```
Ah I see it was fixed here. Please squash this change with the previous commit

gandro replied:
Oh, good catch! Fixed, thanks!

christarazi reviewed (changes requested):
I didn't review the low-level logic and mostly stayed at a high-level. Just a few minor nits to clean up, otherwise the code looks good to me, and is nicely documented :smile:!

christarazi reviewed (approved):
LGTM, doesn't require another review from me on the last nit.
